### DISCUSSION TITLE
VGA_MEM_FIX

### DISCUSSION THREAD

**Discussion by vruppert** [bochs-emu/Bochs#267] [General]
I have now applied a set of fixes for the VGA memory management, but I left them disabled until we have made some more tests. It should fix the Debian / GRUB issue (#257) with both Bochs VBE and Cirrus. If you have some old DOS games, please uncomment the VGA_MEM_FIX define in vgacore.h to test the new code. When we have no regression the new code could be enabled for the planned release 2.8.

**Comment by Vort** [comment]
This is how [Jazz Jackrabbit](https://github.com/bochs-emu/Bochs/files/13539388/Jazz.zip) looks with `VGA_MEM_FIX`:
![bochs_jazz2](https://github.com/bochs-emu/Bochs/assets/1242858/97d344ca-4a02-47c1-b627-d136abb0950e)

**Comment by vruppert** [comment]
Now I found out what's going on. The game uses modeX (mode 13h with chain4 disabled). When playing the game it switches between planar4 writes and odd/even writes with "chain odd/even" turned off. The newly implemented odd/even mode works correctly in text mode, but it has "chain odd/even" turned on. The other mode is not yet implemented. I'll have a look how to handle the address remapping correctly.

**Comment by vruppert** [comment]
Applied a temporary fix for this case. Since the planar mode was used before I added odd/even mode, falling back to planar mode should be okay for now.

**Comment by Vort** [comment]
Ok, Jazz Jackrabbit looks good now.
Next problems are with text mode in [Commander Keen 5](https://github.com/bochs-emu/Bochs/files/13197169/keen5.zip):
![image](https://github.com/bochs-emu/Bochs/assets/1242858/dbd3ea87-4288-4166-a653-da487f28f877)

**Comment by Vort** [comment]
Another problem: Windows XP start to produce crash with Cirrus card and `VGA_MEM_FIX`:
`cursor_address` is `0` here and `bx_gui->cursor_address` is `65535`. `size` of the buffer is `32768`:
```text
>	bochs.exe!bx_gui_c::text_update_common(unsigned char * old_text, unsigned char * new_text, unsigned short cursor_address, bx_vga_tminfo_t * tm_info) Line 1237	C++
 	bochs.exe!bx_vgacore_c::update() Line 1749	C++
 	bochs.exe!bx_svga_cirrus_c::update() Line 1219	C++
 	bochs.exe!bx_vgacore_c::vga_timer_handler(void * this_ptr) Line 2547	C++
 	bochs.exe!bx_virt_timer_c::periodic(unsigned __int64 time_passed, bool mode) Line 165	C++
 	bochs.exe!bx_virt_timer_c::advance_virtual_time(unsigned __int64 time_passed, bool mode) Line 323	C++
 	bochs.exe!bx_virt_timer_c::timer_handler(bool mode) Line 435	C++
 	bochs.exe!bx_virt_timer_c::pc_system_timer_handler_0(void * this_ptr) Line 536	C++
 	bochs.exe!bx_pc_system_c::countdownEvent() Line 382	C++
 	bochs.exe!bx_pc_system_c::tickn(unsigned int n) Line 117	C++
 	bochs.exe!BX_CPU_C::cpu_loop() Line 125	C++
 	bochs.exe!bx_begin_simulation(int argc, char * * argv) Line 1051	C++
 	bochs.exe!bx_real_sim_c::begin_simulation(int argc, char * * argv) Line 930	C++
 	bochs.exe!win32_ci_callback(void * userdata, ci_command_t command) Line 778	C++
 	bochs.exe!bx_real_sim_c::configuration_interface(const char * ignore, ci_command_t command) Line 922	C++
 	bochs.exe!bxmain() Line 336	C++
 	bochs.exe!main(int argc, char * * argv) Line 555	C++
```

**Reply by Vort** [reply]
Fixed in 0e5800bf98162597d856a6a0bb25e3e29cb8027c.

**Reply by vruppert** [reply]
This one should be fixed now. For the Commander Keen issue I need to do some more investigation, The startup screen uses contradictory text mode settings. The new code uses the graphics controller setting for the odd/even mode, but the related setting of the setting of the sequencer is different. I need to find out how to handle this.

**Reply by vruppert** [reply]
Update: I just tried Commander Keen with Qemu, since most of the new code is based on it's implementation. It shows the same issues as Bochs with VGA_MEM_FIX. The DOSBox display looks correct, but I don't really understand it's code. It seems to use the "chain odd/even" bit only. I finally had a look at the PCem source, but I didn't test it yet. It uses the other two odd/even related switches - one for the read case and the other for the write case. If odd/even mode is detected it's way, it rotates the address bits by two bits. I think we have to find the right way ourselves after some more testing.

**Reply by Vort** [reply]
Another source to look at is VirtualBox:
https://www.virtualbox.org/browser/vbox/trunk/src/VBox/Devices/Graphics/DevVGA.cpp

**Reply by vruppert** [reply]
Thanks for pointing this out. I have now applied some new code based on the VirtualBox implementation. It seems to be a fixed version of the Qemu code. I have not yet added the address shift by two instead of one, since I have not yet tested it with Debian / GRUB. The startup screen of Commander Keen is now okay, but the game itself has issues that need to be observed. Jazz Jackrabbit also looks okay after this change.

**Reply by Vort** [reply]
Several more games I tested had no additional problems.
Scroll problems with non-`0` `update_freq` persists, but they are not related to recent changes.

**Comment by Vort** [comment]
One more problem is with [Dangerous Dave in the Haunted Mansion](https://github.com/bochs-emu/Bochs/files/14237223/dave2.zip), image flickers:

https://github.com/bochs-emu/Bochs/assets/1242858/030a13ed-d42c-422c-b317-5d3edb7c62cc

**Reply by vruppert** [reply]
I have currently no idea what can go wrong here. It looks like a timing issue that appears only with VGA_MEM_FIX. It is a planar VGA mode and there seem to be no writes in other modes. When I add BX_INFOs in mem_write() to find out what's going on the flicker doesn't seem to happen.

**Reply by Vort** [reply]
Game stopped glitching like that after I changed `vgamem_mask` from `BX_VGA_THIS s.memsize - 1` to `0x3FFFF`.

**Reply by vruppert** [reply]
I have now added a variable that will be modified in VBE mode for 4-bpp support.
